### DISCUSSION TITLE
Please add support for encryption of DuckDB database file (password protected)

### DISCUSSION THREAD

**Discussion by philipmoore-8451** [duckdb/duckdb#4512] [Ideas]
It would be cool if DuckDB supported encryption of the database file (in the case of native DuckDB format databases).  The duckdb executable (or library) would require a password upon opening of the database file to be able to use the database.

One could also set an environment variable (such as: DUCKDB_DATABASE_PASSWORD for example) so that the executable/library would implicitly open/un-encrypt the database for usage.

SQLite currently licenses this feature separately from their core product (see: https://www.hwaci.com/sw/sqlite/see.html)

This would make it much more secure to be able to pass around DuckDB database files that are natively protected, or encrypted at rest.

Thank you!

**Comment by xland** [comment]
Yes.
It is a important feather for a desktop app with DuckDB.

**Comment by l1t1** [comment]
encrypting big databases  is a heavy operate

**Comment by hannes** [comment]
We're considering adding this as a separately licensed extension indeed.

**Reply by alexandroid11** [reply]
We interested in licensed extension for encryption of DuckDB database file. Do you have it available? Can you provide any details, cost, license details. Thanks.

**Reply by hannes** [reply]
Please ping me at hannes@duckdblabs.com

**Reply by xubobbs** [reply]
does this function provide or not today?

**Comment by yabirgb** [comment]
It would be great if this could be done as it is really useful for local apps

**Comment by colaborado** [comment]
That is also a very big requirement for me. I can't use a database without access protection in any of my projects.

**Comment by ashokgudur** [comment]
I would like and willing to implement this feature just like SQLite did. I’m currently studying it.

**Comment by timothyklim** [comment]
Also, there a fork of sqlite with encryption for free: https://github.com/sqlcipher/sqlcipher

**Reply by mash-graz** [reply]
It's also a free feature of tursos improved sqlite fork `libsql`.

https://docs.turso.tech/libsql#encryption-at-rest

**Comment by hannes** [comment]
This is now planned for the first half of 2025

**Comment by ErenKrt** [comment]
Is there any updates about encryption? @hannes

**Reply by hannes** [reply]
See above? We're still in that first half

**Reply by FrancoisLepoutre** [reply]
Hi Hannes,
 
No hurry indeed! 

...But now that the duckdb format has stabilized - a significant milestone, for sure - some users will appreciate to hear that encryption will be supported. As far as I am concerned, I have no hurry. Timing is not the issue as we can certainly live with current parquet encryption support for a while. But I am curious and would be glad to hear about the coming offer. Both in terms of licensing and in terms of specifications. 

Once you have some specifications in mind, I expect quite a number of users will pay attention to your posts on the subject. One question that we all may have concerning this coming functionality: "What kind of penalty in terms of performance?".

**Reply by llersch** [reply]
Also consider making it FIPS (or equivalent) compliant :)

**Comment by szarnyasg** [comment]
Update: there has been significant work on this, see e.g. https://github.com/duckdb/duckdb/pull/17275 and followup PRs. These features will likely land in DuckDB 1.4 (see https://duckdb.org/cal for release dates).

**Reply by FrancoisLepoutre** [reply]
Hi Gabor. Thanks for this great piece of news! The feature set is great and the tentative timing is also appreciable. They will certainly help all of those who intend to use the duckdb format as a document format. Having some feed back on the cpu cost is also quite useful.
François

**Reply by ErenKrt** [reply]
This news made my day better

**Reply by LuciferSam8086** [reply]
Good news, indeed. This feature will be available as a commercial plugin or directly in DuckDB ?

**Reply by Mytherin** [reply]
It will be available directly in DuckDB.

**Comment by Mytherin** [comment]
This has been implemented and will land in v1.4.0

**Reply by FrancoisLepoutre** [reply]
Hi Mark. I can remember your great talk about the duckdb 1.0 format stabilization. You took the time to explain both the roadmap and the development process. Here we are! For the few of us that use or will use our beloved duck format as a save-and-distribute format, that's a great piece of news.

Thanks to you and the whole team!
